### ISSUE TITLE
Requesting RNG XLA frontend.

### ISSUE THREAD

**Issue by antoniojkim** [google/jax#4555]
I have a particular use case where I would like to convert a model built using a Jax backend to XLA. However, for something like the dropout layer in the model, I am unable to generated the desired XLA.

As can be seen here:
```python
>>> key = jax.random.PRNGKey(1234)
>>> d = np.random.random((10,))
>>> print(jax.xla_computation(lambda d: jax.random.bernoulli(key, 0.5, d.shape) * d)(d).as_hlo_text())
HloModule xla_computation__lambda___2.6

ENTRY xla_computation__lambda___2.6 {
  constant.3 = pred[] constant(false)
  constant.1 = f32[10]{0} constant({0, 0, 1, 0, 0, 1, 0, 0, 1, 0})
  parameter.2 = f32[10]{0} parameter(0)
  multiply.4 = f32[10]{0} multiply(constant.1, parameter.2)
  ROOT tuple.5 = (f32[10]{0}) tuple(multiply.4)
}
```

The bernoulli random distribution generation becomes an array of constants in the XLA computation. However, with something like dropout, I want the random array to be generated each time the forward pass occurs (Note, the above example is a trivialization of the behaviour I want to achieve. I understand that with Jax random number generation I would want to split the key every time I generate numbers). I've searched through the documentation and the repository, but could not find anything that would help me to achieve this.

Looking at the XLA documentation, I see that there are two XLA ops [RngNormal](https://www.tensorflow.org/xla/operation_semantics#rngnormal) and [RngUniform](https://www.tensorflow.org/xla/operation_semantics#rnguniform) that look to perform the desired action. However, as far as I can tell, these ops are not supported by Jax.

Is there a way to achieve my desired behavior using the current Jax library? If not, can someone please create a ticket to support these XLA ops?

**Comment by jakevdp** [comment]
If I understand correctly, I think you can achieve what you want by passing the `key` as a parameter to the function:
```python
key = jax.random.PRNGKey(1234)
d = np.random.random((10,))
print(jax.xla_computation(lambda key, d: jax.random.bernoulli(key, 0.5, d.shape) * d)(key, d).as_hlo_text())
```

**Comment by antoniojkim** [comment]
While this technically does work, and the generated array is no longer a constant in the XLA, the XLA itself is very messy and appears to include all of the ops required to perform the three fry computation that Jax uses to generate random numbers.

It would be so much nicer if Jax supported the XLA RNG ops. Any chance that this could happen sometime in the near future?

**Comment by jakevdp** [comment]
Oh, I see. I'm not sure what the route would be to supporting that. One thing that JAX strives for is invariance between compiled and non-compiled codepaths, and this would be difficult to achieve for general RNG operations in XLA. For example `RngUniform` and `RngNormal` appear to be stateful generators, whereas JAX expressions by design have no state.

It might be possible to lower some jax random code to XLA's `RngBitGenerator`, since it purports to support `rng_three_fry` as a PRNG algorithm, although it looks like XLA uses a 64-bit algorithm while JAX uses 32-bit.

In any case the idea is not something I've heard discussed at all, but perhaps other JAX team members have thoughts?

**Comment by mattjj** [comment]
The main issue is that XLA RNG operations are stateful and so they don't work well with JAX transformations; see the [JAX PRNG design doc](https://github.com/google/jax/blob/master/design_notes/prng.md). You can actually call e.g. [`lax.rng_uniform`](https://github.com/google/jax/blob/d1ca3b3dbe5cf546d6df48c9eb6d766b6a17a66a/jax/lax/lax.py#L5800-L5810) to generate XLA RNG ops (we used them for [MLPerf](https://cloud.google.com/blog/products/ai-machine-learning/google-breaks-ai-performance-records-in-mlperf-with-worlds-fastest-training-supercomputer)), but I don't recommend it because their statefulness will cause transformed code to behave in hard-to-predict ways.

XLA's ThreeFry implementation was added much later, and IIRC doesn't yet do the things we'd need from it (like work the same across backends, maybe? I forget, maybe Peter remembers).

There are probably ways to make the generated HLO appear as a single op, but so far that hasn't been a high priority. Can you say more about why you need that?

**Comment by antoniojkim** [comment]
I'm in the process of breaking down some TensorFlow code and as you may know the TensorFlow repo is very convoluted and full of fluff. So, to try and understand how the TensorFlow model is executing its forward pass, I wanted to compare the XLA generated by TensorFlow with the XLA generated by what I think is the equivalent Jax code and try to reproduce the results of the TensorFlow model using Jax. I've been mostly successful in doing this except for when its comes to the dropout layer which I cannot currently reproduce because the TensorFlow code appears to use the XLA `RngUniform` op, whereas Jax does not.

I understand that if this is not a high priority, I probably won't see it implemented in the Jax library anytime soon. However, if there is a way to have Jax generate the `RngUniform` op when translating the Jax model to XLA, that would be awesome to see.

**Comment by mattjj** [comment]
Maybe for experimentation purposes you should try out `lax.rng_uniform`, noting the warning in the docstring (that the API may go away at any time) and my warning above (that stateful RNG is dangerous with function transformations). At least it'll give you simpler HLO to look at!

Interested if @hawkinsp has any thoughts too.

**Comment by antoniojkim** [comment]
I'll keep the warning in mind. Though, it doesn't look like `lax.rng_uniform` generates what I want anyways:

```python
>>> print(jax.xla_computation(lambda d: lax.rng_uniform(0, 2, d.shape) * d)(d).as_hlo_text())
HloModule xla_computation__lambda___6.6

ENTRY xla_computation__lambda___6.6 {
  constant.3 = pred[] constant(false)
  constant.1 = f32[10]{0} constant({0, 0, 0, 1, 1, 0, 1, 1, 1, 0})
  parameter.2 = f32[10]{0} parameter(0)
  multiply.4 = f32[10]{0} multiply(constant.1, parameter.2)
  ROOT tuple.5 = (f32[10]{0}) tuple(multiply.4)
}
```

Still open to any other suggestions. But I'll go forward with the understanding that this is not something that Jax currently supports.

**Comment by hawkinsp** [comment]
Are you synced to Github head? For your example I get:

```
In [9]: print(jax.xla_computation(lambda d: lax.rng_uniform(0, 2, d.shape) * d)(d).as_hlo_text())
HloModule xla_computation__lambda_.9

ENTRY xla_computation__lambda_.9 {
  constant.2 = pred[] constant(false)
  constant.3 = s32[] constant(0)
  constant.4 = s32[] constant(2)
  rng.5 = s32[2] rng(constant.3, constant.4), distribution=rng_uniform
  convert.6 = f32[2] convert(rng.5)
  parameter.1 = f32[2]{0} parameter(0)
  multiply.7 = f32[2] multiply(convert.6, parameter.1)
  ROOT tuple.8 = (f32[2]) tuple(multiply.7)
}
```

**Comment by mattjj** [comment]
Yeah @antoniojkim that looks like pre-0.2.0 behavior. You should be able to just `pip install --upgrade jax` rather than having to sync to GitHub.

**Comment by antoniojkim** [comment]
Awesome! `jax==0.2.1` appears to behave exactly how I wanted it to. Thank you all for the quick responses!
